compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1809,12 +1811,24 @@ private BType getEffectiveMappingType(BLangRecordLiteral recordLiteral, BType ap
         recordType.tsymbol = recordSymbol;
 
         BLangRecordTypeNode recordTypeNode = TypeDefBuilderHelper.createRecordTypeNode(recordType, pkgID, symTable,
-                                                                                       recordLiteral.pos);
+                                                                                       pos);
         recordTypeNode.initFunction = TypeDefBuilderHelper.createInitFunctionForRecordType(recordTypeNode, env,
                                                                                            names, symTable);
         TypeDefBuilderHelper.addTypeDefinition(recordType, recordSymbol, recordTypeNode, env);
 
-        if (applicableMappingType.tag == TypeTags.MAP) {
+        if (applicableMappingType.tag == TypeTags.RECORD) {
+            BRecordType applicableRecordType = (BRecordType) applicableMappingType;
```

Comment:
Since it is not possible to call record field default closures individually atm, the fix done for the time-being is adding the original type as an included type to make its init method (that initiates the fields) get called.

However this will cause issues for something like 

```ballerina
type Foo record {
    string[] x = ["foo", "bar"];
};

// Panics now with `invalid value for record field 'x': expected value of type 'string[] & readonly', found 'string[]'`,
// since `x` is initialized using the mutable value from `Foo` before the mapping constructor's value.
Foo _ = {readonly x: [""]};
```

This does not panic in Beta3/Beta4 but will panic with this fix.

Related to https://github.com/ballerina-platform/ballerina-lang/issues/24077 in a way.

```ballerina
type Foo record {
    int[] x = [];
};

Foo & readonly _ = {};
```

Also panics at runtime atm.

But if we don't fix this, defaults won't be used in cases like

```ballerina
import ballerina/io;

type Foo record {
    int i;
    string s = "default";
};

public function main() {
    Foo f = {readonly i: 1};
    io:println(f); // `{"i":1,"s":null}` and not `{"i":1,"s":"default"}`
}
```

@hasithaa, thoughts please?

Reply:
+1 for now. will @chiranSachintha fix for closures address this? I guess not.

Reply:
We need to introduce similar closures for record default values, right? Created https://github.com/ballerina-platform/ballerina-lang/issues/33771 to track fixing it. https://github.com/ballerina-platform/ballerina-lang/issues/32012 also depends on it.

Created https://github.com/ballerina-platform/ballerina-lang/issues/33773 to track fixing this specific case.